sdk/aqueduct/integrations/s3_integration.py

```diff
@@ -118,18 +126,33 @@ def file(
                         ArtifactMetadata(
                             id=output_artifact_id,
                             name=artifact_name_from_op_name(op_name),
-                            type=ArtifactType.UNTYPED,
+                            type=artifact_type,
                         ),
                     ],
                 )
             ],
         )
 
-        # Issue preview request since this is an eager execution
-        artifact = artifact_utils.preview_artifact(self._dag, output_artifact_id)
-        self._dag.must_get_artifact(output_artifact_id).type = artifact.type()
+        if execution_mode == ExecutionMode.EAGER:
+            # Issue preview request since this is an eager execution.
+            artifact = artifact_utils.preview_artifact(self._dag, output_artifact_id)
+            if artifact.type() != artifact_type:
+                raise InvalidArtifactTypeException(
+                    "The computed artifact is expected to be type %s, but has type %s"
+                    % (artifact_type, artifact.type())
+                )
 
-        return artifact
+            return artifact
+        else:
+            # We are in lazy mode.
+            if artifact_type == ArtifactType.TABLE:
```

Comment:
Just for my own learning, why this checking is specific to s3? It appears to be something generalizable to all artifacts

Reply:
For S3, we require the user to provide a type hint in the `file` API, so even in lazy mode we can try to use the type hint to create type-specific artifact classes. In SQL, since the type has to be Table, in lazy mode we can also "hardcode" the returned type to be Table, but we don't need to worry about cases like numeric, bool, etc. For generic operators (`@op`), since the output type can be anything and there's no hint from the user, we can only return a `GenericArtifact` with its type being `UNTYPED`.